### PR TITLE
Add support for OTP-22

### LINKED ISSUE #211: Add support for OTP 22
Erlang OTP 22 is out and my system automatically updated with it when I updated all my packages. That promptly broke my ability to compile any code using rusler. Rustler simply issues a panic that its not supported and aborts.

### PR DESCRIPTION
**Pull request by filmor**
Fixes #211.

### CODE REVIEW

File: erl_nif_sys/gen_api.erl [hunk 1]
```diff
@@ -33,8 +33,15 @@ version_opts("2.13") -> [{major,2}, {minor,13}, exception, getenv, time,   % erl
 version_opts("2.14") -> [{major,2}, {minor,14}, exception, getenv, time,   % erlang 21.0
                         dirty_scheduler_opt, nif_2_11, nif_2_12, nif_2_13,
                         nif_2_14];
-version_opts(_) ->
-    io:format("Unsupported Erlang version.\n\nIs the erlang_nif-sys version up to date in the Cargo.toml?\nDoes 'cargo update' fix it?\nIf not please report at https://github.com/goertzenator/erlang_nif-sys.\n"),
+version_opts("2.15") -> [{major,2}, {minor,14}, exception, getenv, time,   % erlang 22.0
```
Comment (chrismcg) [inline erl_nif_sys/gen_api.erl]: Shouldn't this be `{minor,15}`?

Comment (filmor) [reply erl_nif_sys/gen_api.erl]: It should indeed, thanks :)

**Comment by filmor** [comment]
I'll have a look into the failing tests, most of them seem to relate to an issue with copied binaries.

**Comment by filmor** [comment]
Actually, I see those exact same panics on master, and they have been there forever, so the issue was not introduced in this PR. Are those intentionally being ignored?

I'll have a look at the failing resource cleanup test.

**Comment by filmor** [comment]
Just for future reference, I haven't bisected it, but I think the responsible merge is this one: https://github.com/erlang/otp/commit/bc3de715d1aaf6604cd8bde43900f22a631ce548, a fix for bug https://bugs.erlang.org/browse/ERL-863, in particular commit https://github.com/erlang/otp/commit/7a0a52a9a1adc8f87365936f4e76739b7850f1b3, which changes how resource destructors are scheduled.

File: erl_nif_sys/gen_api.erl [hunk 2]
```diff
@@ -33,8 +33,15 @@ version_opts("2.13") -> [{major,2}, {minor,13}, exception, getenv, time,   % erl
...
                         dirty_scheduler_opt, nif_2_11, nif_2_12, nif_2_13,
                         nif_2_14];
-version_opts(_) ->
-    io:format("Unsupported Erlang version.\n\nIs the erlang_nif-sys version up to date in the Cargo.toml?\nDoes 'cargo update' fix it?\nIf not please report at https://github.com/goertzenator/erlang_nif-sys.\n"),
+version_opts("2.15") -> [{major,2}, {minor,15}, exception, getenv, time,   % erlang 22.0
+                        dirty_scheduler_opt, nif_2_11, nif_2_12, nif_2_13,
+                        nif_2_14, nif_2_15];
+version_opts(Ver) ->
+    io:format(
+        "This OTP release uses the unsupported Erlang NIF version ~p.\n\n"
```
Comment (scrogson) [inline erl_nif_sys/gen_api.erl]: `s/the/an`

This OTP release uses an unsupported...

File: erl_nif_sys/gen_api.erl [hunk 3]
```diff
@@ -344,11 +351,24 @@ api_list(Opts) -> [
...
 
-            {"c_int",  "enif_make_map_from_arrays", "env: *mut ErlNifEnv, keys: *const ERL_NIF_TERM, values: *const ERL_NIF_TERM, cnt: usize, map_out: *mut ERL_NIF_TERM"}
+            {"c_int", "enif_make_map_from_arrays", "env: *mut ErlNifEnv, keys: *const ERL_NIF_TERM, values: *const ERL_NIF_TERM, cnt: usize, map_out: *mut ERL_NIF_TERM"}
         ];
         false -> []
-    end.
+    end ++
+    case proplists:get_bool(nif_2_15, Opts) of
+        true -> [
+            {"ErlNifTermType", "enif_term_type", "env: *mut ErlNifEnv, term: *const ERL_NIF_TERM"},
```
Comment (scrogson) [inline erl_nif_sys/gen_api.erl]: This is nice!

File: erl_nif_sys/gen_api.erl (hunk 2 above)
Comment (filmor) [reply erl_nif_sys/gen_api.erl]: Well, the message contains the version, so it reads `This OTP release uses the unsupported Erlang NIF version 2.15` ;)

**Comment by hansihe** [comment]
Thanks!

**Pull request merged**